codalab/lib/bundle_store.py

```diff
@@ -161,28 +226,188 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+        self.ring.remove_node(partition)
+        bundles_to_move = reduce(lambda dirs, files: dirs + files, path_util.ls(old_mdata))
+
+        for bundle in bundles_to_move:
+            new_partition = self.ring.get_node(bundle)
+            relocations[bundle] = os.path.join(self.partitions, new_partition)
+
+        # Copy all bundles off of the old partition to temp directories on the new partition 
+        for bundle, partition in relocations.iteritems():
+            # temporary directory on the partition
+            temp_dir = os.path.join(partition, MultiDiskBundleStore.TEMP_SUBDIRECTORY)
+            from_path = os.path.join(old_mdata, bundle)
+            to_path = os.path.join(temp_dir, 'stage-%s' % bundle)
+            path_util.copy(from_path, to_path)
+
+        # Now that each bundle is on the proper partition, move each from the staging area to the
+        # production mdata/ subdirectory on its partition.
+        for bundle, partition in relocations.iteritems():
+            temp_dir = os.path.join(partition, MultiDiskBundleStore.TEMP_SUBDIRECTORY)
+            from_path = os.path.join(temp_dir, 'stage-%s' % bundle)
+            to_path = os.path.join(partition, MultiDiskBundleStore.DATA_SUBDIRECTORY, bundle)
+            path_util.rename(from_path, to_path)
+
+        # Remove data from partition and unlink from CodaLab
+        print >> sys.stderr, "Cleaning bundles off of partition..."
+        path_util.remove(old_mdata)
+        path_util.remove(old_mtemp)
+        print >> sys.stderr, "Unlinking partition %s from CodaLab deployment..." % partition
+        path_util.remove(partition_abs_path)
+        print >> sys.stderr, "Partition removed successfully from bundle store pool"
+
+    def ls_partitions(self):
+        """List all partitions available for storing bundles and how many bundles are currently stored."""
+        partitions, _ = path_util.ls(self.partitions)
+        print '%d %s' % (len(partitions), 'partition' if len(partitions) == 1 else 'partitions')
+        for d in partitions:
+            partition_path = os.path.join(self.partitions, d)
+            real_path = os.readlink(partition_path)
+            bundles = reduce(lambda x,y: x+y, path_util.ls(os.path.join(partition_path, MultiDiskBundleStore.DATA_SUBDIRECTORY)))
+            print '- %-016s\n\tmountpoint: %s\n\t%d %s' % (d, real_path, len(bundles), 'bundle' if len(bundles) == 1 else 'bundles')
+
     def cleanup(self, model, uuid, except_bundle_uuids, dry_run):
         '''
-        If the given data hash is not needed by any bundle (not in
-        except_bundle_uuids), delete the data.
+        Remove the bundle with given UUID from on-disk storage.
         '''
-        absolute_path = self.get_location(uuid)
-        print >>sys.stderr, "cleanup: data %s" % absolute_path
-        if not dry_run:
-            path_util.remove(absolute_path)
+        try:
+            bundle = model.get_bundle(uuid)
```

Comment:
Though now that I think about it, this would also go ahead and delete bundles that are still running and so their data_hash is also None, so I also need to check bundle state and make sure it's not marked as running?